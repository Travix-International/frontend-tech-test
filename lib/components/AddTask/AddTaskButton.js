import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import styles from './AddTaskButton.scss';
import * as filterActions from '../../actions/filterActions';
import * as taskActions from '../../actions/taskActions';

class AddTaskButton extends Component {
  static propTypes = {
    actions: PropTypes.object.isRequired,
    showCompleted: PropTypes.bool.isRequired
  };

  handleChange = () => {
    this.props.actions.toggleShowCompleted();
  };

  handleAddTaskClick = () => {
    this.props.actions.toggleTaskModal();
  };

  render() {
    const { showCompleted } = this.props;
    const checkboxStyles = `checkbox ${
      showCompleted ? 'checkbox--checked' : ''
    }`;
    return (
      <section className={styles.root}>
        <div className={styles.container}>
          <div className={styles.filters}>
            <div className={checkboxStyles}>
              <input
                checked={showCompleted}
                id="show-completed"
                name="show-completed"
                onChange={this.handleChange}
                type="checkbox"
              />
              <label htmlFor="show-completed">Show completed</label>
            </div>
          </div>
          <button className={styles.button} onClick={this.handleAddTaskClick}>
            Add Task
          </button>
        </div>
      </section>
    );
  }
}

const mapStateToProps = ({ showCompleted }) => ({ showCompleted });

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(
    {
      ...filterActions,
      ...taskActions
    },
    dispatch
  )
});

export default connect(mapStateToProps, mapDispatchToProps)(AddTaskButton);
