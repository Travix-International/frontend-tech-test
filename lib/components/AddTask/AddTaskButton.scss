@import '../../scss/_settings.scss';
@import '../../scss/_objects.scss';

.filters {
  width: 60%;
}

.button {
  @extend .button;
  color: $color-brand;
}

.remaining {
  color: $color-secondary;
  font-weight: bold;
}

.root {
  background: $color-light;
  box-shadow: 0 -5px 5px -5px $color-secondary;
  bottom: 0;
  left: 0;
  padding: 20px 0;
  position: fixed;
  width: 100%;
  z-index: 1;
}

.container {
  align-items: center;
  display: flex;
  justify-content: space-between;
  margin: 0 auto;
  width: 90%;
}

.cancel {
  background: none;
  border: 0;
  font-size: 2.25em;
  color: $color-light;
}

@media (min-width: 768px) {
  .root {
    border-top: 0;
    box-shadow: 0 0;
    background: none;
    position: static;
  }

  .container {
    width: 100%;
  }
}
