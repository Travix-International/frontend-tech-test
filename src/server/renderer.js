import React from 'react';
import ReactServer from 'react-dom/server';
import { StaticRouter as Router, matchPath } from 'react-router';
import { Provider } from 'react-redux';
import { Helmet } from 'react-helmet';
import { /* get as _get, */ isFunction as _isFunction } from 'lodash';
import renderFullPage from './layout';
import configureStore from '../configureStore';
import { entryRouteComponentMap, getEntryAndRoute, entryReducerMap } from './entryAndRoute';

function applyInitStore(req) {
  const entryRouteInfo = getEntryAndRoute(req.path);
  const currentEntry = entryRouteInfo.entry;
  const rootReducer = entryReducerMap[currentEntry];
  const store = configureStore({}, rootReducer);
  req.reduxStore = store;
  req.entryRouteInfo = entryRouteInfo;
  // next();
}

function applyRouteCheckResult(req) {
  const { path, reduxStore: store, query, entryRouteInfo } = req;
  const promises = [];
  const currentEntry = entryRouteInfo.entry;
  const route = entryRouteInfo.route;
  const match = matchPath(path, route);
  const { loadData } = route;

  // set currentEntry to req
  req.currentEntry = currentEntry;

  // it's possible no loadData set
  if (_isFunction(loadData)) {
    let action = loadData(match, query);
    if (action) {
      if (_isFunction(action) || action.type) {
        action = store.dispatch(action);
        if (action.then) {
          promises.push(action);
        }
      }
    }
  }

  return Promise.all(promises);
}

function responsePage(req, res, clientStats) {
  const { url, reduxStore: store, currentEntry } = req;
  const routerContext = {};
  const reduxStateString = JSON.stringify(store.getState()).replace(/</g, '\\u003c');

  const RouteComponent = entryRouteComponentMap[currentEntry];

  const content = (
    <Provider store={store}>
      <Router context={routerContext} location={url}>
        <RouteComponent />
      </Router>
    </Provider>
  );

  // render to sting to get helmet setting
  const contentString = ReactServer.renderToString(content);
  const head = Helmet.renderStatic();
  const htmlString = renderFullPage(contentString, reduxStateString, head, currentEntry, clientStats);

  res.send(htmlString);
}

export default function renderer({ clientStats }) {
  return (req, res) => {
    applyInitStore(req);
    return applyRouteCheckResult(req).then(() => {
      return responsePage(req, res, clientStats);
    });
  };
}
