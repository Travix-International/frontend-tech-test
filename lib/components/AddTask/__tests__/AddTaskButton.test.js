import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import { AddTaskButton } from '../AddTaskButton';

const testProps = {
  actions: {},
  showCompleted: true
};

describe('AddTaskButton', () => {
  let wrapper;
  const mockToggleShowCompletedFn = jest.fn();
  const mockToggleAddModalFn = jest.fn();

  testProps.actions.toggleShowCompleted = mockToggleShowCompletedFn;
  testProps.actions.toggleAddModal = mockToggleAddModalFn;

  const AddTaskButtonComponent = <AddTaskButton {...testProps} />;

  beforeEach(() => {
    wrapper = shallow(AddTaskButtonComponent);
  });

  it('renders correctly', () => {
    const tree = renderer.create(<AddTaskButton {...testProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls toggle modal function when `Add Task` is clicked', () => {
    wrapper.find('.button').simulate('click');
    expect(mockToggleAddModalFn).toBeCalled();
  });

  it('toggles show completed tasks status when checkbox is checked/unchecked', () => {
    wrapper
      .find(`#show-completed`)
      .simulate('change', { target: { checked: false } });
    expect(mockToggleShowCompletedFn).toBeCalled();
  });
});
