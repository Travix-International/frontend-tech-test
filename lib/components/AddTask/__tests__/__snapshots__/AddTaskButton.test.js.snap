// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTaskButton renders correctly 1`] = `
<section
  className="root"
>
  <div
    className="container"
  >
    <div
      className="filters"
    >
      <div
        className="checkbox checkbox--checked"
      >
        <input
          checked={true}
          id="show-completed"
          name="show-completed"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          htmlFor="show-completed"
        >
          Show completed
        </label>
      </div>
    </div>
    <button
      className="button"
      onClick={[Function]}
    >
      Add Task
    </button>
  </div>
</section>
`;
